File: 1136.parallel_courses.py
"""
Approach:

Traditional DFS, returning depth (or inf if cycle detected)

"""

class Solution:
    todo = 0
    pend = 1
    done = 2

    def dfs_visit(self, edge_list, status, depths, node):
        depth = 1
        status[node] = self.pend
        for next_node in edge_list[node]:
            if status[next_node] == self.todo:
                self.dfs_visit(edge_list, status, depths, next_node)
            elif status[next_node] == self.pend:
                depth = float('inf')
            else: # done
                pass
            depth = max(depth, depths[next_node] + 1)
        status[node] = self.done
        depths[node] = depth

    def minimumSemesters(self, n: int, relations: List[List[int]]) -> int:
        edge_list = collections.defaultdict(list)
        for prev_course, next_course in relations:
            edge_list[next_course - 1].append(prev_course - 1)
        
        status = [self.todo] * n
        depths = [0] * n
        for node in range(n):
            if status[node] == self.todo:
                self.dfs_visit(edge_list, status, depths, node)
        
        max_depth = max(depths)
        if max_depth == float('inf'):
            return -1
        return max_depth

File: 1136.parallel_courses.cc
class Solution {
public:
    enum Status {
      TODO = 0,
      PEND = 1,
      DONE = 2
    };
    using Graph = vector<vector<int>>;
    int _max_depth(int node, vector<Status> &status, vector<int> &depths, Graph &graph) {
      assert(node >= 0 && node < status.size());
      if (status[node] == PEND) return INT_MAX;
      if (status[node] == DONE) return depths[node];
      status[node] = PEND;
      int depth = 0;
      for (auto nbor: graph[node]) {
        auto this_depth = _max_depth(nbor, status, depths, graph);
        depth = max(this_depth, depth);
      }
      status[node] = DONE;
      depths[node] = (depth == INT_MAX) ? INT_MAX : depth + 1;
      return depths[node];
    }

    int minimumSemesters(int n, vector<vector<int>>& relations) {
      int max_depth = 0;
      vector<Status> status(n, TODO);
      vector<int> depths(n, 0);
      Graph graph(n, vector<int>());
      for (auto &rel : relations) {
        graph[rel[0]-1].push_back(rel[1]-1);
      }
      for (int node = 0; node != n; node++) {
        int depth = _max_depth(node, status, depths, graph);
        max_depth = max(max_depth, depth);
        if (max_depth == INT_MAX) return -1;
      }
      return max_depth;
    }
};

/*
want to do a topo sort.  this is traditionally done by maintaining a global
dictionary of states: todo, pending, done  for each node in the graph.
we traverse using DFS, transitioning any node from todo->pending upon visiting,
then when returning from the node, from pending->done.

Each node experiences the two transitions exactly once.

Since the leaves don't depend on anything, they are added to the schedule first.

We want to return minimum depth, or -1 if cycle is detected.


*/
